### ISSUE TITLE
[P2] Probe training with interventions

### ISSUE THREAD

**Issue by frankaging** [frankaging/pyvene#26]
Description:
The training for interventions is done through backprop from the model loss. There are other use cases such as intervention itself could be supervised by other objectives such as probes attached to the intervention site.

We can add a basic classification head on top of the intervention object so that additional gradients can be backprop through the probe to the interventions. This could allow this library to support basic probing experiments with desired class labels.

**Comment by aryamanarora** [comment]
For some recent experiments, I used this dummy intervention class to collect activations and train a probe:

```python
class CollectActivation(Intervention):
    
    """Collect activations."""
    def __init__(self, embed_dim, **kwargs):
        super().__init__()
        self.interchange_dim = None
        self.subspace_partition = kwargs["subspace_partition"] \
            if "subspace_partition" in kwargs else None
        self.stored_base = None
        self.stored_src = None
        
    def set_interchange_dim(self, interchange_dim):
        self.interchange_dim = interchange_dim

    def forward(self, base, source):
        self.stored_base = base
        self.stored_src = source
        return base
    
    def get_stored_vals(self):
        return self.stored_base, self.stored_src
    
    def __str__(self):
        return f"CollectActivation()"
```

One idea is to implement probes as iteratively-learned dummy interventions, e.g. for a linear probe we store all the parameters in the intervention object and run a backward pass every time we do a forward pass of the model?

**Comment by frankaging** [comment]
this looks great - i think something like this + generalizing to use cases such as intervening then probing would be the great first step. e.g., this collect activation can be combined with other intervention, like we intervene on layer 1, and collect activations at layer 2, or more nested.
